modules/gateway/rpc.go

```diff
@@ -93,17 +98,34 @@ func (g *Gateway) RegisterConnectCall(name string, fn modules.RPCFunc) {
 // corresponding RPCFunc from g.initRPCs. Future connections to peers will not
 // trigger the RPC to be called on them.
 func (g *Gateway) UnregisterConnectCall(name string) {
-	id := g.mu.Lock()
-	defer g.mu.Unlock(id)
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	if _, ok := g.initRPCs[name]; !ok {
 		build.Critical("ConnectCall not registered: " + name)
 	}
 	delete(g.initRPCs, name)
 }
 
-// listenPeer listens for new streams on a peer connection and serves them via
+// threadedListenPeer listens for new streams on a peer connection and serves them via
 // threadedHandleConn.
-func (g *Gateway) listenPeer(p *peer) {
+func (g *Gateway) threadedListenPeer(p *peer) {
+	if g.threads.Add() != nil {
+		return
+	}
+	defer g.threads.Done()
+
+	// Disconnect from the peer when the stop signal is received.
+	go func() {
+		if g.threads.Add() != nil {
+			return
+		}
+		defer g.threads.Done()
+		<-g.threads.StopChan()
+		if err := p.sess.Close(); err != nil {
```

Comment:
This code has a hidden bug too. `Add` is called inside the goroutine, which means that this goroutine might only start after the gateway has been closed, preventing the peers from disconnecting. This might be the cause of the deadlock. Closing the sessions in `gateway.Close` is easier to implement bug free and is a lot cleaner.